urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/nn/genetic/OptimizableNeuralNetwork.java

```diff
@@ -0,0 +1,313 @@
+import com.lagodiuk.nn.NeuralNetwork;
+import com.lagodiuk.nn.Neuron;
+import com.lagodiuk.nn.ThresholdFunction;
+
+@XmlRootElement
+public class OptimizableNeuralNetwork extends NeuralNetwork implements Chromosome<OptimizableNeuralNetwork>, Cloneable {
+
+	private static double weightsMutationInterval = 1;
+
+	private static double neuronParamsMutationInterval = 1;
+
+	@XmlTransient
+	private Random random = new Random();
+
+	public OptimizableNeuralNetwork() {
+		// Required by JAXB
+	}
+
+	public OptimizableNeuralNetwork(int numberOfNeurons) {
+		super(numberOfNeurons);
+	}
+
+	public OptimizableNeuralNetwork(NeuralNetwork nn) {
+		this.activationIterations = nn.getActivationIterations();
+		this.neurons = nn.getNeurons();
+		this.neuronsLinks = nn.getNeuronsLinks();
+	}
+
+	@Override
+	public List<OptimizableNeuralNetwork> crossover(OptimizableNeuralNetwork anotherChromosome) {
+		OptimizableNeuralNetwork anotherClone = anotherChromosome.clone();
+		OptimizableNeuralNetwork thisClone = this.clone();
+
+		switch (this.random.nextInt(4)) {
+			case 0: {
+				List<Double> thisWeights = thisClone.neuronsLinks.getAllWeights();
+				List<Double> anotherWeights = anotherClone.neuronsLinks.getAllWeights();
+				this.twoPointsWeightsCrossover(thisWeights, anotherWeights);
+				thisClone.neuronsLinks.setAllWeights(thisWeights);
+				anotherClone.neuronsLinks.setAllWeights(anotherWeights);
+			}
+				break;
+			case 1: {
+				List<Double> thisWeights = thisClone.neuronsLinks.getAllWeights();
+				List<Double> anotherWeights = anotherClone.neuronsLinks.getAllWeights();
+				this.uniformelyDistributedWeightsCrossover(thisWeights, anotherWeights);
+				thisClone.neuronsLinks.setAllWeights(thisWeights);
+				anotherClone.neuronsLinks.setAllWeights(anotherWeights);
+			}
+				break;
+			case 2: {
+				this.twoPointsNeuronsCrossover(thisClone.neurons, anotherClone.neurons);
+			}
+				break;
+			case 3: {
+				this.uniformelyDistributedNeuronsCrossover(thisClone.neurons, anotherClone.neurons);
+			}
+				break;
+		// TODO
+		// case 4: {
+		// this.activationIterations += this.random.nextInt(2) -
+		// this.random.nextInt(2);
+		// this.activationIterations = (this.activationIterations < 1) ? 1 :
+		// this.activationIterations;
+		// }
+		// break;
+		}
+
+		List<OptimizableNeuralNetwork> ret = new ArrayList<OptimizableNeuralNetwork>();
+		ret.add(anotherClone);
+		ret.add(thisClone);
+		ret.add(anotherClone.mutate());
+		ret.add(thisClone.mutate());
+		return ret;
+	}
+
+	private void twoPointsWeightsCrossover(List<Double> thisWeights, List<Double> anotherWeights) {
+		int left = this.random.nextInt(thisWeights.size());
+		int right = this.random.nextInt(thisWeights.size());
+		if (left > right) {
+			int tmp = right;
+			right = left;
+			left = tmp;
+		}
+		for (int i = left; i < right; i++) {
+			double thisWeight = anotherWeights.get(i);
+			thisWeights.set(i, anotherWeights.get(i));
+			anotherWeights.set(i, thisWeight);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 106:117, 144:155.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.